dynamodb/schema.go

```diff
@@ -0,0 +1,367 @@
+			tables = append(tables, *t)
+		}
+
+		if result.LastEvaluatedTableName == nil {
+			return tables, nil
+		}
+		input.ExclusiveStartTableName = result.LastEvaluatedTableName
+	}
+}
+
+func processTable(conv *internal.Conv, client dynamoClient, table string, sampleSize int64) error {
+	dySchema := dynamoDBSchema{TableName: table}
+	err := dySchema.analyzeMetadata(client)
+	if err != nil {
+		return err
+	}
+	stats, count, err := dySchema.scanSampleData(client, sampleSize)
+	if err != nil {
+		return err
+	}
+	dySchema.inferDataTypes(stats, count)
+	conv.SrcSchema[table] = dySchema.genericSchema()
+	return nil
+}
+
+type dynamoDBSchema struct {
+	TableName   string
+	ColumnNames []string
+	ColumnTypes map[string]colType
+	PrimaryKeys []string
+	SecIndexes  []index
+}
+
+type colType struct {
+	Name     string
+	Nullable bool
+}
+
+type index struct {
+	Name string
+	Keys []string
+}
+
+func (s *dynamoDBSchema) analyzeMetadata(client dynamoClient) error {
+	input := &dynamodb.DescribeTableInput{
+		TableName: aws.String(s.TableName),
+	}
+
+	result, err := client.DescribeTable(input)
+	if err != nil {
+		return fmt.Errorf("failed to make a DescribeTable API call for table %v: %v", s.TableName, err)
+	}
+
+	// Primary keys
+	for _, i := range result.Table.KeySchema {
+		s.PrimaryKeys = append(s.PrimaryKeys, *i.AttributeName)
+	}
+
+	// Secondary indexes
+	for _, i := range result.Table.GlobalSecondaryIndexes {
+		var keys []string
+		for _, j := range i.KeySchema {
+			keys = append(keys, *j.AttributeName)
+		}
+		s.SecIndexes = append(s.SecIndexes, index{Name: *i.IndexName, Keys: keys})
+	}
+
+	return nil
+}
+
+func (s *dynamoDBSchema) scanSampleData(client dynamoClient, sampleSize int64) (map[string]map[string]int64, int64, error) {
+	// A map from column name to a count map of possible data types.
+	stats := make(map[string]map[string]int64)
+	var count int64
+	// Build the query input parameters
+	params := &dynamodb.ScanInput{
+		TableName: aws.String(s.TableName),
+	}
+
+	for {
+		// Make the DynamoDB Query API call
+		result, err := client.Scan(params)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to make Query API call for table %v: %v", s.TableName, err)
+		}
+
+		// Iterate the items returned
+		for _, attrsMap := range result.Items {
+			for attrName, attr := range attrsMap {
+				if _, ok := stats[attrName]; !ok {
+					stats[attrName] = make(map[string]int64)
+				}
+				incTypeCount(attrName, attr, stats[attrName])
+			}
+
+			count++
+			if count >= sampleSize {
+				return stats, count, nil
+			}
+		}
+		if result.LastEvaluatedKey == nil {
+			break
+		}
+		// If there are more rows, then continue.
+		params.ExclusiveStartKey = result.LastEvaluatedKey
+	}
+	return stats, count, nil
+}
+
+func incTypeCount(attrName string, attr *dynamodb.AttributeValue, s map[string]int64) {
+	switch {
+	case attr.S != nil:
+		s[typeString]++
+	case attr.BOOL != nil:
+		s[typeBool]++
+	case attr.N != nil:
+		if numericParsable(*attr.N) {
```

Comment:
Maybe leave a comment about our treatment of numeric and possibilities for future modifications?

Something like:

// We map the DynamoDB Number type into Spanner's NUMERIC type
// if it fits and STRING otherwise. Note that DyanamoDB's Number
// type has more precision/range than Spanner's NUMERIC.
// We could potentially do a more detailed analysis and see if
// the number fits in an INT64 or FLOAT64, but we've chosen to
// keep the analysis simple for the moment.